nemo/collections/tts/modules/tacotron2.py

```diff
@@ -0,0 +1,568 @@
+        decoder_inputs = self.parse_decoder_inputs(decoder_inputs)
+        decoder_inputs = torch.cat((decoder_input, decoder_inputs), dim=0)
+        decoder_inputs = self.prenet(decoder_inputs)
+
+        self.initialize_decoder_states(memory, mask=~get_mask_from_lengths(memory_lengths))
+
+        mel_outputs, gate_outputs, alignments = [], [], []
+        while len(mel_outputs) < decoder_inputs.size(0) - 1:
+            decoder_input = decoder_inputs[len(mel_outputs)]
+            mel_output, gate_output, attention_weights = self.decode(decoder_input)
+
+            mel_outputs += [mel_output.squeeze(1)]
+            gate_outputs += [gate_output.squeeze()]
+            alignments += [attention_weights]
+
+        mel_outputs, gate_outputs, alignments = self.parse_decoder_outputs(mel_outputs, gate_outputs, alignments)
+        return mel_outputs, gate_outputs, alignments
+
+    def infer(self, *, memory, memory_lengths):
+        """ Decoder inference
+        PARAMS
+        ------
+        memory: Encoder outputs
+        RETURNS
+        -------
+        mel_outputs: mel outputs from the decoder
+        gate_outputs: gate outputs from the decoder
+        alignments: sequence of attention weights from the decoder
+        """
+        decoder_input = self.get_go_frame(memory)
+
+        if memory.size(0) > 1:
+            mask = ~get_mask_from_lengths(memory_lengths)
+        else:
+            mask = None
+
+        self.initialize_decoder_states(memory, mask=mask)
+
+        mel_lengths = torch.zeros([memory.size(0)], dtype=torch.int32)
+        not_finished = torch.ones([memory.size(0)], dtype=torch.int32)
+        if torch.cuda.is_available():
+            mel_lengths = mel_lengths.cuda()
+            not_finished = not_finished.cuda()
+
+        mel_outputs, gate_outputs, alignments = [], [], []
+        stepped = False
+        while True:
+            decoder_input = self.prenet(decoder_input, inference=True)
+            mel_output, gate_output, alignment = self.decode(decoder_input)
+
+            dec = torch.le(torch.sigmoid(gate_output.data), self.gate_threshold).to(torch.int32).squeeze(1)
+
+            not_finished = not_finished * dec
+            mel_lengths += not_finished
+
+            if self.early_stopping and torch.sum(not_finished) == 0 and stepped:
+                break
+            stepped = True
+
+            mel_outputs += [mel_output.squeeze(1)]
+            gate_outputs += [gate_output]
+            alignments += [alignment]
+
+            if len(mel_outputs) == self.max_decoder_steps:
+                logging.warning("Reached max decoder steps %d.", self.max_decoder_steps)
+                break
+
+            decoder_input = mel_output
+
+        mel_outputs, gate_outputs, alignments = self.parse_decoder_outputs(mel_outputs, gate_outputs, alignments)
+
+        return mel_outputs, gate_outputs, alignments, mel_lengths
+
+    def save_to(self, save_path: str):
```

Comment:
I think save_to/restore_from can be removed now

Reply:
Done in next commit

Reply:
Undone